packages/favorites/src/FavoritesContext.tsx

```diff
@@ -0,0 +1,128 @@
+    onMutate({ activeMutationId }) {
+      setActiveMutationId(activeMutationId);
+    },
+    onSettled() {
+      setActiveMutationId(null);
+    },
+  });
+
+  useEffect(() => {
+    avMessages.subscribe(AV_INTERNAL_GLOBALS.FAVORITES_CHANGED, (messagesData) => {
+      queryClient.setQueryData('favorites', messagesData.favorites || []);
+    });
+
+    return () => avMessages.unsubscribe(AV_INTERNAL_GLOBALS.FAVORITES_CHANGED);
+  }, [queryClient]);
+
+  const deleteFavorite = async (id: string) => {
+    if (favorites) {
+      const response = await submitFavorites({
+        favorites: favorites.filter((favorite: Favorite) => favorite.id !== id),
+        activeMutationId: id,
+      });
+
+      sendUpdateMessage(response.favorites);
+    }
+  };
+
+  const addFavorite = async (id: string) => {
+    if (!favorites) return false;
+
+    if (favorites.length >= MAX_FAVORITES) {
+      openMaxModal();
+      return false;
+    }
+
+    const maxFavorite = favorites.reduce<Favorite | null>((accum, fave) => {
+      if (!accum || fave.pos > accum.pos) {
+        accum = fave;
+      }
+      return accum;
+    }, null);
+
+    setActiveMutationId(id);
+
+    const newFavPos = maxFavorite ? maxFavorite.pos + 1 : 0;
+    const response = await submitFavorites({
+      favorites: [...favorites, { id, pos: newFavPos }],
+      activeMutationId: id,
+    });
+
+    sendUpdateMessage(response.favorites);
+
+    const isFavorited = response.favorites.find((f) => f.id === id);
+
+    return !!isFavorited;
+  };
+
+  return (
+    <FavoritesContext.Provider
+      value={{
+        favorites,
+        status,
+        isLoading,
+        activeMutationId,
+        deleteFavorite,
+        addFavorite,
+      }}
+    >
+      {children}
+    </FavoritesContext.Provider>
+  );
+};
+
+export const useFavorites = (
+  id: string
+): {
+  isFavorited: boolean;
+  status: StatusUnion;
+  isDisabled: boolean;
+  isActiveMutation: boolean;
+  toggleFavorite: () => Promise<void>;
+  allFavorites?: Favorite[];
+} => {
+  const context = useContext(FavoritesContext);
+
+  if (context === null) {
+    throw new Error('useFavorites must be used within a FavoritesProvider');
+  }
+  const { favorites, status, activeMutationId, deleteFavorite, addFavorite } = context;
+
+  const isActiveMutation = activeMutationId === id;
+
+  const isFavorited = useMemo(() => {
+    const fav = favorites?.find((f) => f.id === id);
+    return !!fav;
+  }, [favorites, id]);
+
+  const toggleFavorite = async () => (isFavorited ? deleteFavorite(id) : addFavorite(id));
+
+  const isDisabled = status === 'loading' || status === 'idle' || activeMutationId !== null;
+
+  return {
+    isFavorited,
+    status,
+    // try to put spinner on the button that was clicked, and subtly disable the other buttons
+    isDisabled,
+    isActiveMutation,
+    toggleFavorite: isDisabled ? asyncNoOp : toggleFavorite,
+    allFavorites: favorites,
```

Comment:
Why `allFavorites` and not `favorites` here?

Reply:
This was originally meant to be temporary. I wanted to render a list of all my favorites, and I realized react-query already had that data. I don't have a good reason why I called it 'allFavorites'. I guess just because the rest of the hook centered around data for the specific favorite for the id passed in. 

I'm inclined to just take this out. Do you see any value in returning all the favorites from this hook? If it does have value, I think instead we should just provide a separate `useFavoritesQuery` export, which would just be the useQuery for favorites that I already have.

What do you think?

Reply:
I think that's a good idea to have a generic one that returns _all_ of them, and then a more specific hook that handles the specific logic

Reply:
Turns out I was already exporting `useFavoritesQuery` from the package. So I'm just removing the 'allFavorites' property here. I'm not currently using that property from the useFavorites hook, and I'm not sure how much use there is for it outside of its internal use within the package.